_posts/2024-08-07-accessing-kafka-with-gateway-api.md

```diff
@@ -0,0 +1,675 @@
+
+```
+apiVersion: gateway.networking.k8s.io/v1
+kind: GatewayClass
+metadata:
+  name: my-gateway-class
+  namespace: envoy-gateway-system
+spec:
+  controllerName: gateway.envoyproxy.io/gatewayclass-controller
+  parametersRef:
+    group: gateway.envoyproxy.io
+    kind: EnvoyProxy
+    name: my-proxy-config
+    namespace: envoy-gateway-system
+---
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: EnvoyProxy
+metadata:
+  name: my-proxy-config
+  namespace: envoy-gateway-system
+spec:
+  provider:
+    type: Kubernetes
+    kubernetes:
+      envoyDeployment:
+        replicas: 1
+      envoyService:
+        type: NodePort
+        patch:
+          value:
+            spec:
+              ports:
+              # Port 9092 on your laptop gets forwarded to NodePort 30992 on the KIND cluster.
+              - name: kafka-port
+                nodePort: 30992
+                port: 9092
+                protocol: TCP
+                targetPort: 9092
+---
+```
+If you're following along in my GitHub:
+```
+kubectl apply -f gateway-class.yaml
+```
+
+Now that we have a `GatewayClass` which is properly configured by an Envoy Gateway `EnvoyProxy` resource, we can create our `Gateway`. As discussed earlier in the setup overview, we need want the Envoy Proxy pods to listen on port `9092`. That listener will be configured to have Passthrough TLS.
+
+```
+apiVersion: gateway.networking.k8s.io/v1
+kind: Gateway
+metadata:
+  name: my-gateway
+  # Envoy Gateway can control gateways in all namespaces. Depending on your
+  # cluster permissions model, a Gateway can be in the envoy-gateway-system namespace
+  # or in the same namespace as your applications.
+  namespace: strimzi
+spec:
+  gatewayClassName: my-gateway-class
+  listeners:
+  - name: kafka-listener
+    protocol: TLS
+    port: 9092
+    tls:
+      mode: Passthrough
+```
+
+In the github repo:
+
+```
+kubectl apply -f gateway.yaml
+```
+
+This will create a `Gateway` resource, and the Envoy Gateway controller will deploy an Envoy pod in the `envoy-gateway-system` namespace to handle traffic for that gateway class. Once that is done, you should see some pods in the `envoy-gateway-system` namespace, like the following:
+
+```
+-> kubectl get pods --namespace envoy-gateway-system
+NAME                                                 READY   STATUS    RESTARTS   AGE
+envoy-strimzi-my-gateway-1c7c06f0-5446c7ff7b-vpd6m   1/2     Running   0          22s
+envoy-gateway-8595cc9fbc-2bjn5                       1/1     Running   0          96s
+```
+
+The second pod is the Envoy Gateway controller, which reconciles all Gateway API-related resources. The first `Pod` was created by the controller to route all traffic for the `my-gateway` `Gateway` which we created in the `strimzi` namespace.
```

Comment:
So ... all this big part about deploying kind, deploying envoy, talking about mapping 9092 local to 30992 pod and then to 9092 envoy pod etc etc ... is simple to be confused and lost for someone who has never used kind first, gateway API and envoy implementation. I think a diagram with the port mapping across pods, where the gateway controller stands, the envoy pod, etc etc would be really helpful.

Reply:
Roger roger, will add one today. Thanks!